File: kotlin.web.demo.server/static/js/views/InputDialogView.js
/*
 * Copyright 2000-2014 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Created by Semyon.Atamas on 8/27/2014.
 */


var InputDialogView = (function () {

    var dialog = document.createElement("div");
    dialog.id = "input-dialog";

    var text = document.createElement("span");
    text.id = "input-dialog-text";
    dialog.appendChild(text);

    var input = document.createElement("input");
    input.id = "input-dialog-input";
    input.type = "text";
    dialog.appendChild(input);

    var message = document.createElement("div");
    dialog.appendChild(message);
    message.className = "input-dialog-error-message";

    $(dialog).dialog({
        resizable: false,
        modal: "true",
        width: 380,
        autoOpen: false,
        open: function(){
            input.select();
        }
    });


    $(dialog).keydown(function (event) {
        if (event.keyCode == 13) { /*enter*/
            $(this).parent()
                .find("button:eq(1)").trigger("click");
        } else if (event.keyCode == 27) { /*escape*/
            $(this).dialog("close");
        }
        event.stopPropagation();
    });

    function InputDialogView(title, inputText, buttonText) {
        var instance = {

            open: function (callback, defaultValue) {

                input.focus();

                validationResult({valid: true});

                input.oninput = function () {
                    validationResult(instance.validate(input.value));
                };

                if (defaultValue != null && defaultValue != undefined) {
                    var verifiedDefaultValue = defaultValue;
                    var i = 1;
                    while (!instance.validate(verifiedDefaultValue).valid) {
                        verifiedDefaultValue = defaultValue + i;
                        ++i;
                    }
                    input.value = verifiedDefaultValue;
                    //input.select();
                }

                $(dialog).dialog('option', 'title', title);
                text.innerHTML = inputText;
                $(dialog).dialog("option", "buttons", [
                        {
                            text: buttonText,
                            click: function (event) {
                                var validationResult = instance.validate(input.value);
                                if (validationResult.valid) {
                                    callback(input.value);
                                    $(this).dialog("close");
                                } else {
                                    validationResult(validationResult);
                                }
                                event.stopPropagation();
                            }
                        },
                        {
                            text: "Cancel",
                            click: function (event) {
                                $(this).dialog("close");
                                event.stopPropagation();
                            }
                        }
                    ]
                );
                $(dialog).dialog("open");
            },
            validate: function (input) {
                return {valid: true};
            }
        };

        function validationResult(result) {
            input.style.outlineColor = result.valid ? "" : "red";
            $(dialog).parent().find("button:eq(1)").button("option", "disabled", !result.valid);
            if (!result.valid && result.message != null) {
                message.innerHTML = result.message;
            } else {
                message.innerHTML = "";
            }
            input.focus();
        }

        return instance;
    }

    return InputDialogView;
})();

File: kotlin.web.demo.frontend/src/views/dialogs/InputDialogView.kt
/*
 * Copyright 2000-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package views.dialogs

import kotlinx.html.*
import kotlinx.html.dom.create
import kotlinx.html.js.div
import org.w3c.dom.HTMLInputElement
import org.w3c.dom.events.Event
import utils.KeyCode
import utils.jquery.jq
import utils.jquery.ui.Dialog
import utils.jquery.ui.DialogButton
import kotlin.browser.document


open class InputDialogView(
        val dialogTitle: String,
        val inputLabel: String,
        val okButtonCaption: String,
        val defaultValue: String,
        val validate: (String) -> ValidationResult,
        val callback: (UserInput) -> Unit,
        val submitOnEnter: Boolean = true
) {
    private val dialogElement = render()

    private val errorMessageElement = jq(dialogElement).find(".input-dialog-error-message")[0]

    protected val inputElement = jq(dialogElement).find("#input-dialog-input")[0] as HTMLInputElement

    private val dialog: Dialog = Dialog(
            dialogElement,
            resizable = false,
            modal = true,
            width = 380,
            onOpen = { event, ui ->
                inputElement.select()
            },
            onClose = { this.close() },
            buttons = arrayOf(
                    DialogButton(
                            text = okButtonCaption,
                            click = { event: Event ->
                                event.stopPropagation()
                                callback(getUserInput())
                                this.close()
                            }
                    ),
                    DialogButton(
                            text = "Cancel",
                            click = { event: Event ->
                                event.stopPropagation()
                                this.close()
                            }
                    )
            )
    )

    init {
        if(submitOnEnter) {
            jq(dialogElement).keydown({ event ->
                when (event.keyCode) {
                    KeyCode.ENTER.code -> {
                        val okButton = getDialogButton(dialogElement, 1)
                        okButton.trigger("click")
                    }
                }
                event.stopPropagation()
            })
        }

        inputElement.oninput = {
            processValidationResult(validate(inputElement.value))
        }
        inputElement.select()
    }

    open fun render() = document.create.div {
        id = "input-dialog"
        title = dialogTitle
        span {
            +inputLabel
            id = "input-dialog-text"
        }
        input {
            id = "input-dialog-input"
            type = InputType.text
            value = getVerifiedDefaultValue(defaultValue, validate)
        }
        div {
            classes = setOf("input-dialog-error-message")
        }
    }

    protected open fun getUserInput() = UserInput(inputElement.value)

    fun close() {
        dialog.destroy()
        dialogElement.parentNode?.removeChild(dialogElement);
    }

    private fun processValidationResult(result: ValidationResult) {
        val okButton = getDialogButton(dialogElement, 1)
        okButton.button("option", "disabled", !result.valid)
        errorMessageElement.innerHTML = if (result.valid) "" else result.message
        inputElement.style.outlineColor = if (result.valid) "" else "red"
        inputElement.focus()
    }

    protected fun getVerifiedDefaultValue(defaultValue: String, validate: (String) -> ValidationResult): String {
        var verifiedDefaultValue = defaultValue
        var i = 1
        while (!validate(verifiedDefaultValue).valid) {
            verifiedDefaultValue = defaultValue + i
            ++i
        }
        return verifiedDefaultValue
    }
}

open class UserInput(val value: String);

data class ValidationResult(val valid: Boolean, val message: String = "")